contexts/ProposalMarketsContext.tsx

```diff
@@ -0,0 +1,602 @@
+    placeOrderTransactions: (
+        amount: number,
+        price: number,
+        market: MarketAccountWithKey,
+        limitOrder?: boolean | undefined,
+        ask?: boolean | undefined,
+        pass?: boolean | undefined,
+        indexOffset?: number | undefined,
+    ) => Promise<any>;
+    placeOrder: (
+        amount: number,
+        price: number,
+        limitOrder?: boolean,
+        ask?: boolean,
+        pass?: boolean,
+    ) => Promise<void>;
+    cancelOrder: (order: OpenOrdersAccountWithKey) => Promise<void>;
+}
+
+export const ProposalMarketsContext = createContext<ProposalInterface | undefined>(undefined);
+
+export const useProposalMarkets = () => {
+    const context = useContext(ProposalMarketsContext);
+    if (!context) {
+        throw new Error('useProposalMarkets must be used within a ProposalMarketsContextProvider');
+    }
+    return context;
+};
+
+
+
+export function ProposalMarketsProvider({
+    children,
+    proposalNumber,
+    fromProposal,
+}: {
+    children: React.ReactNode;
+    proposalNumber?: number | undefined;
+    fromProposal?: ProposalAccountWithKey;
+}) {
+    const provider = useProvider();
+    const openBookProgram = new Program<OpenbookV2>(OPENBOOK_IDL, OPENBOOK_PROGRAM_ID, provider);
+    const client = useQueryClient();
+    const { openbook, openbookTwap, proposals } =
+        useAutocrat();
+    const { connection } = useConnection();
+    const wallet = useWallet();
+    const sender = useTransactionSender();
+    const { placeOrderTransactions, settleFundsTransactions,
+        cancelOrderTransactions } = useOpenbookTwap();
+    const {
+        program: vaultProgram,
+    } = useConditionalVault();
+    const [loading, setLoading] = useState(false);
+    const [markets, setMarkets] = useState<Markets>();
+    const [orders, setOrders] = useState<OpenOrdersAccountWithKey[]>([]);
+    const [bids, setBids] = useState<any[][]>();
+    const [asks, setAsks] = useState<any[][]>();
+    const [wsConnected, setWsConnected] = useState<boolean>(false);
+    const [spreadString, setSpreadString] = useState<string>();
+    const [lastSlotUpdated, setLastSlotUpdated] = useState<number>();
+    const isPassMarket = true;
+
+    const proposal = useMemo<Proposal | undefined>(
+        () =>
+            proposals?.filter(
+                (t) =>
+                    t.account.number === proposalNumber ||
+                    t.publicKey.toString() === fromProposal?.publicKey.toString(),
+            )[0],
+        [proposals, fromProposal, proposalNumber],
+    );
+
+    const fetchMarketsInfo = useCallback(
+        debounce(async () => {
+            const fetchProposalMarketsInfo = async () => {
+                setLoading(true);
+                if (!proposal || !openbook || !openbookTwap || !openbookTwap.views || !connection) {
+                    return;
+                }
+                const accountInfos = await connection.getMultipleAccountsInfo([
```

Comment:
Once we load this, it doesn't change, in fact once a proposal is live, this never changes. So may be worth breaking this out into much longer term cache / storage in that we effectively never need touch this again. This is where API could come into play, but also something to consider while we're breaking this up into more manageable pieces.

Reply:
We can also lazy load all the other proposals over time too, since a user may get there eventually, but not entirely sure if that's needed, but something to consider.

Reply:
The only thing is maybe the account data changes for say the twap and pass fail and conditional stuff, but if that's so we only need to update THOSE, not everything, which saves on SOME request burden.